Add free text filter to Auto-Type dialog

hifi opened the pull request:
When you have a lot of Auto-Type matches for the same window title, some quick search is helpful.

The input field has default focus and up/down arrow keys are captured which move the selection in the list. To make this a bit better, it would work with multiple words split into separate filters that narrow down the list.

Note: This does not yet pass the code quality requirements. The existing view model filter is pulled with a `qobject_cast<>()` which may be a bit hacky.

[TIP]:  # ( Provide a general summary of your changes in the title above ^^ )

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ New feature (non-breaking change which adds functionality)

## Description and Context
Implements #2944.

## Screenshots
[TIP]:  # ( Do not include screenshots of your actual database! )
![image](https://user-images.githubusercontent.com/106598/55670104-ead2f080-5888-11e9-961c-22e4413ef76a.png)


## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )
Manual tests on Linux.

## Checklist:
[NOTE]: # ( Please go over all the following points. )
[NOTE]: # ( Again, remove any lines which don't apply. )
[NOTE]: # ( Pull Requests that don't fulfill all [REQUIRED] requisites are likely )
[NOTE]: # ( to be sent back to you for correction or will be rejected.  )
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**

hifi commented:
Tested on Windows 10, works fine there as well.

One persisting issue without touching the view model filter is that the pattern column is also searched. If I were to modify that class then I'd also add multiple word filter but I'm unsure if that would break something else.

This could be merged as-is after it passes quality checks and the multi word filter could be another PR in the future.

0pLuS0 commented:
Can someone please explain how this works, I don't get the need for this?

As an example I have a lot of Auto-Type matches for the same window, but the usernames are all different, and I don't see how someone can have several accounts with everything the same, that they are confused and need to do a search...

Hmm

hifi commented:
I feel obligated to answer this since I implemented this feature for _myself_.

I manage AWS accounts in the tens at work and the amount is only going to increase. All of the accounts have a name of "AWS" and their auto-type sequence matches the generic AWS login page. Groups have the name of the customer in question. There is no way to make the auto-type sequence unique.

Compared to going through the list manually searching for _the_ specific entry I need against typing the first few letters of the account/customer name and hitting return is _much_ more convenient and less error prone.

When I implemented this feature I specifically wanted to keep the arrow keys functioning like before so it doesn't break the workflow of anyone not using the new search field.

droidmonkey commented:
This is also the entry point to a feature where you can search your whole database for an Auto-Type entry if there is none to your liking in the dialog.

0pLuS0 commented:
@hifi so is the Title or Username you are talking about called AWS?

The ```Title``` could be; ```AWS1 - AWS2```?

I just assumed if the Title were organized in a specific way, then there would be no need for this...

Hmm
